pkg/agent/multicast/mcast_discovery.go

```diff
@@ -158,47 +206,56 @@ func (s *IGMPSnooper) processPacketIn(pktIn *ofctrl.PacketIn) error {
 	if err != nil {
 		return err
 	}
+	var events []*mcastGroupEvent
+	packetLen := pktIn.Data.Len()
 	switch igmp.GetMessageType() {
 	case protocol.IGMPv1Report:
 		fallthrough
 	case protocol.IGMPv2Report:
 		mgroup := igmp.(*protocol.IGMPv1or2).GroupAddress
 		klog.V(2).InfoS("Received IGMPv1or2 Report message", "group", mgroup.String(), "interface", iface.InterfaceName, "pod", podName)
-		event := &mcastGroupEvent{
+		events = []*mcastGroupEvent{{
 			group: mgroup,
 			eType: groupJoin,
 			time:  now,
 			iface: iface,
-		}
-		s.validatePacketAndNotify(event)
+		}}
 	case protocol.IGMPv3Report:
 		msg := igmp.(*protocol.IGMPv3MembershipReport)
-		for _, gr := range msg.GroupRecords {
+		events = make([]*mcastGroupEvent, len(msg.GroupRecords))
+		for i, gr := range msg.GroupRecords {
 			mgroup := gr.MulticastAddress
 			klog.V(2).InfoS("Received IGMPv3 Report message", "group", mgroup.String(), "interface", iface.InterfaceName, "pod", podName, "recordType", gr.Type, "sourceCount", gr.NumberOfSources)
 			evtType := groupJoin
 			if (gr.Type == protocol.IGMPIsIn || gr.Type == protocol.IGMPToIn) && gr.NumberOfSources == 0 {
 				evtType = groupLeave
 			}
-			event := &mcastGroupEvent{
+			events[i] = &mcastGroupEvent{
 				group: mgroup,
 				eType: evtType,
 				time:  now,
 				iface: iface,
 			}
-			s.validatePacketAndNotify(event)
 		}
-
 	case protocol.IGMPv2LeaveGroup:
 		mgroup := igmp.(*protocol.IGMPv1or2).GroupAddress
 		klog.V(2).InfoS("Received IGMPv2 Leave message", "group", mgroup.String(), "interface", iface.InterfaceName, "pod", podName)
-		event := &mcastGroupEvent{
+		events = []*mcastGroupEvent{{
 			group: mgroup,
 			eType: groupLeave,
 			time:  now,
 			iface: iface,
+		}}
+	}
+	for i := range events {
+		events[i].packetLen = uint64(packetLen)
```

Comment:
It seems event doesn't really need to have `packetLen` field, it's just for saving a parameter of `validate`? maybe just pass `pktIn.Data` to `validate`?

Reply:
updated